Kernel/Graphics/FramebufferDevice.cpp

```diff
@@ -0,0 +1,156 @@
+    VERIFY(!m_userspace_real_framebuffer_vmobject);
+
+    auto vmobject = AnonymousVMObject::create_for_physical_range(m_framebuffer_address, page_round_up(framebuffer_size_in_bytes()));
+    if (!vmobject)
+        return ENOMEM;
+    m_userspace_real_framebuffer_vmobject = vmobject;
+
+    auto result = process.space().allocate_region_with_vmobject(
+        range,
+        vmobject.release_nonnull(),
+        0,
+        "Framebuffer",
+        prot,
+        shared);
+    if (!result.is_error()) {
+        m_userspace_framebuffer_region = result.value();
+        dbgln("{} {} {}", m_userspace_framebuffer_region->size(), m_real_framebuffer_region->size(), m_swapped_framebuffer_region->size());
+    }
+    return result;
+}
+
+void FramebufferDevice::dectivate_writes()
+{
+    ScopedSpinLock lock(m_activation_lock);
+    if (!m_userspace_framebuffer_region)
+        return;
+    memcpy(m_swapped_framebuffer_region->vaddr().as_ptr(), m_real_framebuffer_region->vaddr().as_ptr(), page_round_up(framebuffer_size_in_bytes()));
+    auto vmobject = m_swapped_framebuffer_vmobject;
+    m_userspace_framebuffer_region->set_vmobject(vmobject.release_nonnull());
+    m_userspace_framebuffer_region->remap();
+}
+void FramebufferDevice::activate_writes()
+{
+    ScopedSpinLock lock(m_activation_lock);
+    if (!m_userspace_framebuffer_region || !m_real_framebuffer_vmobject)
+        return;
+    // restore the image we had in the void area
+    // FIXME: if we happen to have multiple Framebuffers that are writing to that location
+    // we will experience glitches...
+    memcpy(m_real_framebuffer_region->vaddr().as_ptr(), m_swapped_framebuffer_region->vaddr().as_ptr(), page_round_up(framebuffer_size_in_bytes()));
+    auto vmobject = m_userspace_real_framebuffer_vmobject;
+    m_userspace_framebuffer_region->set_vmobject(vmobject.release_nonnull());
+    m_userspace_framebuffer_region->remap();
+}
+
+String FramebufferDevice::device_name() const
+{
+    return String::formatted("fb{}", minor());
+}
+
+UNMAP_AFTER_INIT void FramebufferDevice::initialize()
+{
+    m_real_framebuffer_vmobject = AnonymousVMObject::create_for_physical_range(m_framebuffer_address, page_round_up(framebuffer_size_in_bytes()));
+    VERIFY(m_real_framebuffer_vmobject);
+    m_real_framebuffer_region = MM.allocate_kernel_region_with_vmobject(*m_real_framebuffer_vmobject, framebuffer_size_in_bytes(), "Framebuffer", Region::Access::Read | Region::Access::Write);
+    VERIFY(m_real_framebuffer_region);
+    m_swapped_framebuffer_vmobject = AnonymousVMObject::create_with_size(page_round_up(framebuffer_size_in_bytes()), AllocationStrategy::AllocateNow);
+    VERIFY(m_swapped_framebuffer_vmobject);
+    m_swapped_framebuffer_region = MM.allocate_kernel_region_with_vmobject(*m_swapped_framebuffer_vmobject, page_round_up(framebuffer_size_in_bytes()), "Framebuffer Swap (Blank)", Region::Access::Read | Region::Access::Write);
+    VERIFY(m_swapped_framebuffer_region);
+}
```

Comment:
Do we need to allocate them into kernel space permenantly?

Reply:
Yes. When we switch between graphical mode and textual mode, we need to ensure we restore everything (especially in graphical mode).